### PR TITLE
Adding LDAP support

### PR DESCRIPTION
**Pull request by Krakinou**
Adding LDAP support:
image

Make it work:
* Add [ldapserver]:[port] as LDAP Provider URL
* Add Distinguished name with "%s" where the user name should be.
* Login with your LDAP username and LDAP password

Calibre-web users must be created first with the same username as per the LDAP database.
As activating the LDAP will shunt the database password, do not activate it in the initial setting but first create an administrator user with same name as the one from the LDAP server.

If python-ldap is not already installed on your server, you need to run 
```pip install --target vendor -r optionnal-requirements-ldap.txt```
On ubuntu, you need to run ```sudo apt-get install libsasl2-dev python-dev libldap2-dev libssl-dev``` to be able to pip install it.

Todo:
 * Translation
 * Documentation

### CODE REVIEW

**Comment by Krakinou** [comment]
Hmm,
I might have an issue regarding the ```import ldap``` command which is not conditioned, and so an error is triggered when running the program if python-ldap was not installed.
Might I put python-ldap in requirements.txt?

**Comment by OzzieIsaacs** [comment]
Please don‘t add it to the requirement.txt, I‘d like to keep the foodprint as small as possible for normal users. It might also take some time to merge it, as I‘d like to complete my testing first, than doing a first release and i‘d Like to find a solution to combine all alternative registration routines (like plug-ins)

**Comment by Fmstrat** [comment]
LDAP is a very common authentication and not a huge footprint. I think there are a lot of users who would rather have this than wait for a full plugin implementation, personally. Granted, I'm biased.

**Comment by teknofile** [comment]
This is (slightly) off-topic for this bug request, but not really a "new" bug and I'm not sure where else to ask as I don't seem to be able to find a mailing list or message group or anything....

Has any work been done to date (even high level PoC, considerations) towards a plugin architecture?

**Comment by OzzieIsaacs** [comment]
Beside the fact I mentioned this idea in some threads, definitly not. My idea was to implement different login providers with Plugins, but found that it looked to complicated after Spendengeldern some Time on thinking on it (If plugins store Data the database has to be extended for every plugin)

**Comment by skyforce97809** [comment]
import ldap
ImportError: No module named ldap

**Comment by Krakinou** [comment]
Did you run `pip install --target vendor -r optionnal-requirements-ldap.txt`?

**Comment by skyforce97809** [comment]
yes i run it
and i had set the URL
and DN
when i applied the settings and  trying to login its giving me
500 page error

**Comment by Krakinou** [comment]
Hi, please do not merge as I spotted some issues (for example, opds based on basic auth will be broken if you use this PR)
I'm working on a revamp based on https://github.com/admiralobvious/flask-simpleldap

**Comment by OzzieIsaacs** [comment]
I already merged it to the development branch. If possible it would be great if you could use this branch for the reworked version.

**Comment by Krakinou** [comment]
Here is the reworked version
Changelog :
- cleaner code
- implement [flask_simpleldap](https://github.com/admiralobvious/flask-simpleldap)
- much more options:
Capture du 2019-06-08 01-08-59
- opds is now working as basic auth is handled
- test if ldap is working before actually activating it

Make it work:
in calibre-web folder:
`pip install --target vendor -r optionnal-requirements-ldap.txt`
Activate ldap and enter configuration in the configuration panel
log out and log in back with your LDAP password
**Users are not yet replicated from the ldap directory to the calibre-web database, so DO NOT activate LDAP if you haven't created first an calibre-web admin user with same username as a ldap one.**

Options:
- Use LDAP Authentication: general LDAP switch
- LDAP server Hostname, port,  : self explanatory
- LDAP schema : ldap or ldaps
- LDAP service administrator user and password are required for  simpleldap to work. This is not yet use but I plan to activate user replication in a new version. **LDAP password is stored encoded in the database but is not hashed/salted or whatever**.
- SSL, TLS and Certificate are not tested, feedback welcome.
- LDAP DN : The distinguished name to use as the search base..
- LDAP user object filter: The filter to use when searching for a user object. Default: ‘(&(objectclass=Person)(userPrincipalName=%s))’
- OpenLDAP : switch on if your using an openldap server.

@OzzieIsaacs:  I'm a little confuse : do you want me to create a new PR to the dev branch?

File: cps/ub.py [hunk 1]
```diff
@@ -703,6 +729,74 @@ def migrate_Database():
...
+        session.query(exists().where(Settings.config_ldap_use_ssl)).scalar()
+    except exc.OperationalError:
+        conn = engine.connect()
+        conn.execute("ALTER TABLE Settings ADD column `config_ldap_use_ssl` INTEGER DEFAULT 0")
+        session.commit()
+    try:
+        session.query(exists().where(Settings.config_ldap_use_tls)).scalar()
+    except exc.OperationalError:
+        conn = engine.connect()
+        conn.execute("ALTER TABLE Settings ADD column `cconfig_ldap_use_tls` INTEGER DEFAULT 0")
```
Comment (Philantrop) [inline cps/ub.py]:
```suggestion
        conn.execute("ALTER TABLE Settings ADD column `config_ldap_use_tls` INTEGER DEFAULT 0")
```

File: cps/ub.py [hunk 2]
```diff
@@ -703,6 +729,74 @@ def migrate_Database():
...
+        session.query(exists().where(Settings.config_ldap_port)).scalar()
+    except exc.OperationalError:
+        conn = engine.connect()
+        conn.execute("ALTER TABLE Settings ADD column `config_ldap_port` INTEGER DEFAULT ''")
+        session.commit()
+    try:
+        session.query(exists().where(Settings.config_ldap_schema)).scalar()
+    except exc.OperationalError:
+        conn = engine.connect()
+        conn.execute("ALTER TABLE Settings ADD column `config_ldap_schema ` String DEFAULT ''")
```
Comment (Philantrop) [inline cps/ub.py]:
```suggestion
        conn.execute("ALTER TABLE Settings ADD column `config_ldap_schema` String DEFAULT ''")
```

**Comment by ykorzikowski** [comment]
ATM not working for me. 

```
[2019-06-08 23:27:11,962] INFO in web: LDAP Login failed for user "admin, IP-address :192.168.1.247
::ffff:172.17.0.1 - - [2019-06-08 23:27:11] "POST /calibre-web/login?next=%2Fcalibre-web%2F HTTP/1.0" 200 5741 0.011149
[2019-06-08 23:27:25,504] INFO in web: Login failed for user "ykorzikowski" IP-adress: 192.168.1.247
::ffff:172.17.0.1 - - [2019-06-08 23:27:25] "POST /calibre-web/login?next=%2Fcalibre-web%2F HTTP/1.0" 200 5741 0.010518
[2019-06-08 23:27:32,935] INFO in web: Login failed for user "yko" IP-adress: 192.168.1.247
::ffff:172.17.0.1 - - [2019-06-08 23:27:32] "POST /calibre-web/login?next=%2Fcalibre-web%2F HTTP/1.0" 200 5741 0.010809
```

User admin is not in the ldap but in the internal user direcotry of ldap. Why isn't the ldap plugin trying to verify my account with the ldap? Or is the log message wrong?

**Comment by OzzieIsaacs** [comment]
> @OzzieIsaacs: I'm a little confuse : do you want me to create a new PR to the dev branch?
@Krakinou, yes I would prefer it, if you could create a PR for the development branch. Your old PR is already merged there.

**Comment by Krakinou** [comment]
Closing as new PR for develop branch #941 
@ykorzikowski : What did you put as DN and user object filter? The example I provided are just example, data depend on your LDAP server settings.

**Comment by qwertydumbfish** [comment]
@Krakinou hi sorry im new at calibre web and linux, trying to get this ldap checkbox to even be in my calibre web but it is simply not there. I have installed calibre web without docker, and idk if that has smth to do with the ldap (i guess not) but the calibre web installed in docker has ldap authentification but even there i cant seem to configure it correctly. 

I tried your solution with:
**pip install --target vendor -r optionnal-requirements-ldap.txt**

but i get this error: 

**Could not open requirements file: [Errno 2] No such file or directory: 'optional-requirements-ldap.txt'**

**Comment by qwertydumbfish** [comment]
@OzzieIsaacs moin, I seem to not even have the checkbox or the opportunity to configure LDAP, what could the issue be? Should I install first LDAP in my Debian? [I have installed calibre web in Debian but without docker] Once again im kinda new in this but I really would like to learn more since I find it very interesting, thanks!

**Comment by OzzieIsaacs** [comment]
The optional-requirements-ldap.txt file is renamed to optional-requirements.txt (Please open the file, search for the ldap section, delete everything else, save the file) and try 
pip install --target vendor -r optionnal-requirements.txt

Please have a look at the documentation in the wiki:https://github.com/janeczku/calibre-web/wiki/LDAP-Login

If you have further problems, please open a new issue.

**Pull request closed without merging**